### DISCUSSION TITLE
Charging Cost at AC Stations in the Charging Cost Stats dashboard include some DC supercharger charges

### DISCUSSION THREAD

**Discussion by tnepveu** [jheredianet/Teslamate-CustomGrafanaDashboards#67] [Q&A] [answered]
The title says all , 

Supercharge, other DC stations charge (only one in my case) and total cost is good but the AC cost stats in off , after a few trials, I discovered that some but not all Tesla Supercharge are included in the Charging Cost at AC Stations :

All Supercharge have a geofence with a name starting with "Tesla Supercharger"

![Capture d’écran 2024-02-16 à 16 48 30](https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/assets/157983484/5219ca1e-a789-482f-bef6-86dd05bcb051)

**Comment by tnepveu** [comment]
my guess is that - may be - some of the supercharge does not have the DC flag (charger_voltage and/or fast_charger_present) correctly set 

So it would be a DB issue rather than a dashboard issue.

As I am totally ignorant of postgresql and sql in general for that matter , I don't know how to check my hypothesis.

Thanks in advance, love your dashboards !

**Comment by jheredianet** [comment]
> Thanks in advance, love your dashboards!
Hi, I'm glad you love the dashboards. 
About your question, for the AC charging costs, it doesn't consider the geofence or names, it's based on the "fast_charger_present" field. 
So, it could be what you say. What model is your car?

If you can run this query in your database, you will get the cost of only AC charging sessions (assuming that your carid is 1, if not change it accordantly. On dates, the same, just considering this year, you can modify it as you wish). 
So, you can have an idea why 

`
WITH data AS (
	SELECT c.charging_process_id 
	FROM charges c
	LEFT JOIN charging_processes cp ON cp.id = c.charging_process_id
	WHERE 
		cp.car_id = 1 AND NOT c.fast_charger_present
        GROUP BY 1
)
SELECT
       coalesce(sum(cp.cost) , 0)
FROM
	charging_processes cp, data
WHERE
  end_date >= '20240101'
  AND cp.cost IS NOT NULL
  AND cp.id IN (data.charging_process_id);
`

Also, you may browse your database to know how is your car logging the DC charges with this query:

`
SELECT * FROM charges 
ORDER BY date DESC
 LIMIT 10;
`

You should get something like this:
![image](https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/assets/45735191/08d5be06-2ea5-445c-a38f-a18f692fb8ab)

**Comment by tnepveu** [comment]
Hello Juan Carlos

Thanks for the suggestion , I managed to browse my DB with your indications and indeed there is an issue.

The last event of supercharges (when I disconnect the CCS plug) counted as AC have fast_charger_present set to FALSE where those counted correctly as DC do not have it.

So if, for some reasons, you disconnect the CCS Plug not cleanly, the last event of the supercharge will not have fast_charger_present set to TRUE and the session will be counted as AC charge

I checked only 2 Supercharge for now , I will check more sessions to be 100% sure

Not sure if I am clear enough

The car is a EU Model 3 RWD FTR

**Reply by jheredianet** [reply]
It's strange, nobody has reported something like this. I have a model 3 standard range plus from August 2020 and does not have that behaviour. I will ask some other friends about this.

**Comment by tnepveu** [comment]
Hello Juan Carlos 

Confirmed , I checked a fair amount of charge in my DB and every supercharge counted as an AC Charge have a least one line where fast_charger_present is set to false

it is always the last event of the supercharge session :

**clean supercharge (last 2 events) :**


 70428 | 2024-02-10 13:42:53.569 | f                 |           100 |               52.48 |                      0 |                |                    16 |             0 |               2 | t                    | IEC               | Tesla              | Combo             |                 421.21 |                          |         10.0 |                 161 | f              |                         |                 421.21 |                  100
 70427 | 2024-02-10 13:42:33.381 | f                 |            99 |               52.48 |                      0 |                |                    16 |             8 |               2 | t                    | IEC               | Tesla              | Combo             |                 419.60 |                          |         10.0 |                 161 | f              |                         |                 419.60 |                   99

**dirty supercharge (last 2 events) :**

 69955 | 2024-02-10 10:07:19.229 | f                 |            70 |               36.90 |                      0 |                |                    16 |             0 |               2 | f                    | <invalid>         | Tesla              | Combo             |                 296.65 |                          |          9.5 |                 160 | f              |                         |                 296.65 |                   70
 69954 | 2024-02-10 10:07:13.98  | f                 |            70 |               36.77 |                      0 |                |                    16 |            68 |               2 | t                    | IEC               | Tesla              | Combo             |                 295.75 |                          |          9.5 |                 160 | f              |                         |                 295.75 |                   70


I can provide you my complete charge table if needed (not sure how to dump it)

in my opinion , unless my car have an issue with its charge port, it looks like this is a fairly common event when disconnecting a CCS Combo2 plug from the car (I am betting it's less common with a NACS plug) . So the dashboard should take this into account to be sure that the cost displayed are accurate

 Or may be be even Teslamate should clean charge datas when logging as obviously , a charge can not be both AC and DC at the same time  ?

Cheers

**Reply by jheredianet** [reply]
As I said before, I cannot reproduce that behaviour on my car. A charge cannot be both AC and DC at the same time.
To be more accurate, maybe you should try these steps:

1. Try to find you last DC charge session ID, from the "Charges Dashboard" as shown in the screen shot below. Just over the pointer on the date and in the status bar you will see the "Charging_Process_ID"

![image](https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/assets/45735191/23f00793-38da-4f10-aecd-6d2a3a331bb6)


2. Then run this query in order to see the 10 last events of thar charge. Replace the XXXX with the id from the previous step.

`
SELECT "date", battery_level, charge_energy_added, fast_charger_present
FROM charges WHERE charging_process_id=XXXX ORDER BY date DESC LIMIT 10
`

![image](https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/assets/45735191/b6f99688-5339-4290-8b20-9696a9ea34f8)

As you can see, in my case, the information it's not wrong. Let me see yours (please try to make a screen shot with the headers of the records, so it's easier to read)

**Comment by stauffenberg2020** [comment, marked as answer]
This has been fixed as of #83

**Comment by tnepveu** [comment]
I confirm that it's now fixed in latest release 👍 
![Capture d’écran 2024-04-27 à 17 12 00](https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/assets/157983484/c8ce03f1-4723-476f-9346-b8cee55a4ae0)

Many thanks for the hard work !
